pkg/zkproofs/zero_balance.go

```diff
@@ -0,0 +1,150 @@
+package zkproofs
+
+import (
+	"crypto/rand"
+	"encoding/json"
+	"errors"
+	"github.com/coinbase/kryptology/pkg/core/curves"
+	"github.com/sei-protocol/sei-cryptography/pkg/encryption/elgamal"
+)
+
+// ZeroBalanceProof represents a zero-knowledge proof that a ciphertext encrypts 0 number.
+type ZeroBalanceProof struct {
+	Yp curves.Point
+	Yd curves.Point
+	Z  curves.Scalar
+}
+
+// NewZeroBalanceProof generates a zero-knowledge proof that a ciphertext encrypts 0 number.
+func NewZeroBalanceProof(
+	keypair *elgamal.KeyPair,
+	ciphertext *elgamal.Ciphertext,
+) (*ZeroBalanceProof, error) {
+	if keypair == nil || keypair.PublicKey == nil || keypair.PrivateKey == nil {
+		return nil, errors.New("keypair is invalid")
+	}
+
+	if ciphertext == nil || ciphertext.D == nil || ciphertext.C == nil {
+		return nil, errors.New("ciphertext is invalid")
+	}
+
+	// Extract necessary values
+	P := keypair.PublicKey
+	s := keypair.PrivateKey
+	D := ciphertext.D
+
+	// Generate random masking factor y
+	y := curves.ED25519().Scalar.Random(rand.Reader)
+
+	// Compute Yp = y * P and Yd = y * D
+	Yp := P.Mul(y)
+	Yd := D.Mul(y)
+
+	// Append commitments to the transcript
+	transcript := NewEqualityProofTranscript()
```

Comment:
Looks like we already use this EqualityProofTranscript here as well - could consider in the future generalize this EqualityProofTranscript method of creating challenges across all our proofs

Reply:
All done